tsdb/agent/storage.go

```diff
@@ -0,0 +1,678 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package agent
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wal"
+	"go.uber.org/atomic"
+)
+
+var (
+	ErrUnsupported = errors.New("unsupported operation with WAL-only storage")
+)
+
+// Default values for options.
+var (
+	DefaultTruncateFrequency = 2 * time.Hour
+	DefaultMinWALTime        = 5 * time.Minute
+	DefaultMaxWALTime        = 4 * time.Hour
```

Comment:
In TSDB we don't have a concept of knowing the unit of time, and we should follow the same here and not have `time.Duration` (and wall clock comparison) with samples and rather have these variables as `int64`. You can take inspiration from [here](https://github.com/prometheus/prometheus/blob/7ff76ed1abadcf36d3a71b17fad80916f2601890/tsdb/db.go#L51).